### PR TITLE
[android] Add website:menu tag support 

### LINKED ISSUE #2376: Add support for website:menu tag
This tag has recently been featured in a mapping action and usage has increased quite a bit:
https://taginfo.openstreetmap.org/keys/website:menu#chronology

It would be pretty cool if OM could display a direct link to the menu of restaurant / bar / club etc.
This would make it feel more useful when browsing such places and removes the part where the user has to click the normal website link and then look for a link to the menu

https://wiki.openstreetmap.org/wiki/Key:website:menu

### PR DESCRIPTION
**Pull request by Arthur-GYT**
All is in the title ^^.

Close #2376

I've tested this PR on this node (`France_Free County_North.mwm`) : https://www.openstreetmap.org/node/6637693465

I leave it to those who wish (and can) to add this to the IOS version. As I don't own a Mac, it's impossible for me to take care of this ^^.

Screenshot_20240508-190830

Screenshot_20240506-165510

### CODE REVIEW

**Comment by Arthur-GYT** [comment]
I can't test it, I don't find (and know) any restaurant with this tag. Any suggestion ?

**Comment by Arthur-GYT** [comment]
> Good feature That's require a map generation, to see this tag in place page

Okay, so I need to wait or can I do something to test it ?

**Comment by Arthur-GYT** [comment]
> Good feature
> That's require a map generation, to see this tag in place page

Okay, I will take a look. Just one more thing, I don't need to commit map generation, right ?

**Comment by RedAuburn** [comment]
> Okay, I will take a look. Just one more thing, I don't need to commit map generation, right ?

nope 👍️

File: android/app/src/main/res/layout/place_page_website_menu.xml
Comment (Jean-BaptisteC) [inline android/app/src/main/res/layout/place_page_website_menu.xml:15]: Why Github icon?

Comment (Arthur-GYT) [reply android/app/src/main/res/layout/place_page_website_menu.xml:15]: For now it's a test. I will change it. But, how icons are designed for organic maps ?

Comment (Jean-BaptisteC) [reply android/app/src/main/res/layout/place_page_website_menu.xml:15]: On UI, we use icons proposed by Google and Apple.
WDYT to you same icons has website tag or [menu icon](https://fonts.google.com/icons?selected=Material%20Symbols%20Outlined%3Amenu_book%3AFILL%400%3Bwght%40400%3BGRAD%400%3Bopsz%4024)

Comment (Arthur-GYT) [reply android/app/src/main/res/layout/place_page_website_menu.xml:15]: This [icon](https://fonts.google.com/icons?selected=Material+Symbols+Outlined:restaurant_menu:FILL@0;wght@400;GRAD@0;opsz@24&icon.size=24&icon.color=%23e8eaed&icon.query=menu) look good

**Comment by Arthur-GYT** [comment]
Ready for review ^^ !!

**Comment by RedAuburn** [comment]
can you post a screenshot of how it looks? :)

File: android/app/src/main/res/layout/fragment_editor.xml
Comment (RedAuburn) [inline android/app/src/main/res/layout/fragment_editor.xml:219]: IMO it'd be better to move this below the Wi-Fi section (`id/block_wifi`), as adding wifi is more useful and likely to do when surveying

**Review by RedAuburn — CHANGES_REQUESTED** [review]
looks good overall! a small nit:

**Comment by Arthur-GYT** [comment]
> can you post a screenshot of how it looks? :)

It's uploaded in my first message ^^.

File: android/app/src/main/res/layout/fragment_editor.xml
Comment (Arthur-GYT) [reply android/app/src/main/res/layout/fragment_editor.xml:219]: Yeah, but in the case of the website:menu is a “sub-tag” of website, I think that putting them in a row would be more coherent than separating them by the wifi tag.

Comment (RedAuburn) [reply android/app/src/main/res/layout/fragment_editor.xml:219]: mm fair enough, seeing it in the screenshot it doesn't look too bad

**Comment by RedAuburn** [comment]
should it show the link to the menu, or would it be better to use the text "Open Menu in Browwer"? 
the menu link in the screenshot looks good, but often it's a messy link to a PDF:

eg. 
https://www.openstreetmap.org/node/10271116595

where the link is (or was at least)
`http://79.170.40.239/devthecoriander.com/wp-content/uploads/2020/01/The-Coriander-Buckhurst-Hill_Essex-take-away-menu-proof.pdf`

**Comment by biodranik** [comment]
1. Cuisine icons and menu icons are very similar and confusing. They should be distinct and clear.
2. As it is a menu web URL, then it should clearly say "Menu website link" or "Menu link" or something similar in the editor.
3. Both cuisine and menu should be nearby in POI (they are closely related) and in the editor (we really want to encourage users to add menus, not wifis or something else, this is very valuable data for planning a visit to a restaurant).
4. In the Place Page it is not clear what does an icon and a website URL mean. I suggest replacing it with a clear text: Check Menu.

**Comment by Arthur-GYT** [comment]
@biodranik I updated the code with your recommendations ^^.

**Comment by Arthur-GYT** [comment]
> should it show the link to the menu, or would it be better to use the text "Open Menu in Browser"? the menu link in the screenshot looks good, but often it's a messy link to a PDF:
> 
> eg. https://www.openstreetmap.org/node/10271116595
> 
> where the link is (or was at least) `http://79.170.40.239/devthecoriander.com/wp-content/uploads/2020/01/The-Coriander-Buckhurst-Hill_Essex-take-away-menu-proof.pdf`

I think it's solved with recent changes ^^.

**Comment by Arthur-GYT** [comment]
There's a build that fails, is it because of my PR or the master branch code?

**Comment by RedAuburn** [comment]
> There's a build that fails for iOS, is it because of my PR or the master branch code (I rebased the branch on master)?

i don't think it's related, but i'll re-run it to double-check 👍️

**Comment by RedAuburn** [comment]
It'd be better to still show the normal website URL as the link text, instead of "Check Website" :) 

"Check Menu" looks good though!

**Comment by Arthur-GYT** [comment]
> It'd be better to keep the normal website URL as the button text, instead of "Check Website" 
> 
> "Check Menu" looks good though! :) 

@biodranik What do you think ?

**Comment by Misalf-git** [comment]
```
🌐︎ Website:  burgers-example.com
📖︎ Menu:     burgers-example.com/tasty-menu
```

**Comment by dvdmrtnz** [comment]
> ```
> 🌐︎ Website:  burgers-example.com
> 📖︎ Menu:     burgers-example.com/tasty-menu
> ```

I like how transparent this is. If the menu url is ugly that’s the restaurant's fault. IMO, I would not hide the url.

**Comment by Arthur-GYT** [comment]
I have 3 opinions, what am I supposed to do ^^'?

**Comment by Arthur-GYT** [comment]
To display the url without it being prepended, perhaps I can set it to be small at the bottom of the “Check Website” button? What do you think @RedAuburn, @biodranik ?

**Comment by Arthur-GYT** [comment]
Well I do this. @biodranik @RedAuburn Let me know what you think ^^.
Screenshot_20240508-190830

**Comment by dvdmrtnz** [comment]
> Well I do this. @biodranik @RedAuburn Let me know what you think ^^. Screenshot_20240508-190830

Looks nice!

**Comment by biodranik** [comment]
What is the value in seeing the URL to the menu? It is often ugly and hosted on some 3party resources for smaller restaurants/cafes.

"Check Menu" can be confused with "The Menu of a Check" or "Check's Menu". Would "View Menu" be less confusing?

**Comment by Arthur-GYT** [comment]
@biodranik 

Imagine someone adds a malicious link on Open Street Map for a restaurant. If we hide the link, the user will have no idea what he's about to open.

And since we can see the website link, it would make sense to have a “call to action” with the link underneath (bearing in mind that, unlike the website, the menu link is reduced in size).

**Comment by Misalf-git** [comment]
> What is the value in seeing the URL to the menu?
The fact that it's an external resource.
Also copy/pasting.

> Would "View Menu" be less confusing?
Yes, indeed.

**Comment by Arthur-GYT** [comment]
Well I change the string to `View Menu`. Tell me what you think ^^.

image

**Comment by biodranik** [comment]
It looks ok, thanks. Is there a way to percent-unescape URLs that are printed to users?

**Comment by Arthur-GYT** [comment]
> It looks ok, thanks. Is there a way to percent-unescape URLs that are printed to users?



> It looks ok, thanks. Is there a way to percent-unescape URLs that are printed to users?

Yeah, for website and website menu or just website menu ?

**Comment by Arthur-GYT** [comment]
@biodranik It's done ^^

image

**Review by biodranik — COMMENTED** [review]
@kirylkaveryn can you please add iOS support too in a similar way in another PR based on this PR (so it can be rebased later if necessary)?

@Arthur-GYT can you please share the generated map file to test your changes?

**Comment by Arthur-GYT** [comment]
@biodranik 

https://we.tl/t-OeMPk4Rq9J

**Comment by Misalf-git** [comment]
`de = Speisekarte`
"Menu" translates to "Speisekarte" (like _food-card_) distinctly.

If I may suggest translation here as the GitHub app must be smarting me out again, not allowing me to comment at `strings.txt` directly.

**Comment by Arthur-GYT** [comment]
> `de = Speisekarte` "Menu" translates to "Speisekarte" (like _food-card_) distinctly.
> 
> If I may suggest translation here as the GitHub app must be smarting me out again, not allowing me to comment at `strings.txt` directly.

So I should replace "de = Menü-Link" by "de = Speisekarte" ? And for the other key ?

**Comment by Misalf-git** [comment]
Are both strings used and do they need to be phrased differently?

```
  [website_menu]
    en = Menu link
    ...
    de = Menü-Link

```

`de = Link zur Speisekarte`

---

```
  [view_menu]
    en = View Menu
    ...
    de = Menü ansehen
```

`de = Speisekarte ansehen`

---

I would translate both strings to just
`de = Speisekarte` if they don't need to differ.

**Comment by Arthur-GYT** [comment]
> Are both strings used and do they need to be phrased differently?
> 
> ```
>   [website_menu]
>     en = Menu link
>     ...
>     de = Menü-Link
> ```
> 
> `de = Link zur Speisekarte`
> 
> ```
>   [view_menu]
>     en = View Menu
>     ...
>     de = Menü ansehen
> ```
> 
> `de = Speisekarte ansehen`
> 
> I would translate both strings to just `de = Speisekarte` if they don't need to differ.

^^'

https://github.com/organicmaps/organicmaps/pull/8052#issuecomment-2095841143

**Comment by Arthur-GYT** [comment]
@biodranik So with the message posted yesterday, I conclude that this PR will wait for the iOs implementation, before being merge?

**Comment by kirylkaveryn** [comment]
@biodranik i need some time to implement this feature in iOS.
Should this PR be merged first or we can wait a little bit?

File: android/app/src/main/java/app/organicmaps/bookmarks/data/MapObject.java
Comment (biodranik) [inline android/app/src/main/java/app/organicmaps/bookmarks/data/MapObject.java:275]: This decoded URL is not saved anywhere in the Editor, right? And it is clickable, correct?

File: android/app/src/main/res/drawable/ic_website_menu.xml
Comment (biodranik) [inline android/app/src/main/res/drawable/ic_website_menu.xml:1]: Icon is from Material icons, right?

File: data/strings/strings.txt
Comment (biodranik) [inline data/strings/strings.txt:30178]:
```suggestion
    comment = Restaurant or other food place's menu URL field in the Editor
```

File: data/strings/strings.txt
Comment (biodranik) [inline data/strings/strings.txt:30180]:
```suggestion
    en = Menu Link
```

File: data/strings/strings.txt
Comment (biodranik) [inline data/strings/strings.txt:30223]:
```suggestion
    comment = Button in the Place Page which opens a restaurant or other food place's menu in a browser
```

File: data/strings/strings.txt
Comment (biodranik) [inline data/strings/strings.txt:30229]:
```suggestion
    be = Паглядзець меню
```

File: data/strings/strings.txt
Comment (biodranik) [inline data/strings/strings.txt:30234]:
```suggestion
    de = Speisekarte
```

In German, there is a separate word for a restaurant menu. Other languages need to distinguish a restaurant menu vs application menu.

File: data/strings/strings.txt
Comment (biodranik) [inline data/strings/strings.txt:30263]:
```suggestion
    uk = Переглянути меню
```

**Review by biodranik — COMMENTED** [review]
Thanks, I'm ok to merge it after a few minor fixes.

@vng PTAL, it would be great to run the generator with these changes.

File: android/app/src/main/res/drawable/ic_website_menu.xml
Comment (Arthur-GYT) [reply android/app/src/main/res/drawable/ic_website_menu.xml:1]: Yeah, why ?

File: android/app/src/main/java/app/organicmaps/bookmarks/data/MapObject.java
Comment (Arthur-GYT) [reply android/app/src/main/java/app/organicmaps/bookmarks/data/MapObject.java:275]: Yes, why ?

File: data/strings/strings.txt
Comment (Arthur-GYT) [reply data/strings/strings.txt:30263]: Done ^^.

File: android/app/src/main/res/drawable/ic_website_menu.xml
Comment (biodranik) [reply android/app/src/main/res/drawable/ic_website_menu.xml:1]: To check if license terms are ok, and if it's mentioned in data/copyright.html

**Review by biodranik — APPROVED** [review]
Thank you!

@vng merge please

**Comment by kirylkaveryn** [comment]
@biodranik @Arthur-GYT in iOS we don't have cells with the subtitles in the PP. But without the subtitle, this field may not be clear. Should we add a new cell (it may not be easy because of the current PP implementation) or use the same style as wiki/fb/link?

**Comment by biodranik** [comment]
@kirylkaveryn Let's make it simpler, by not showing the URL, and improve later if necessary.

**Comment by kirylkaveryn** [comment]
@Arthur-GYT @biodranik should the new type be validated in the `indexer/editable_map_object.cpp`? :
```cpp
bool EditableMapObject::IsValidMetadata(MetadataID type, std::string const & val
{
  switch (type)
  {
  case MetadataID::FMD_WEBSITE: return ValidateWebsite(value);
  case MetadataID::FMD_WEBSITE_MENU: return ValidateWebsite(value); // <--- ?

void EditableMapObject::SetMetadata(MetadataID type, std::string value)
{
  switch (type)
  {
  case MetadataID::FMD_WEBSITE: value = ValidateAndFormat_website(value); break;
  case MetadataID::FMD_WEBSITE_MENU: value = ValidateAndFormat_website(value); break;  // <--- ?
...

```

**Comment by Arthur-GYT** [comment]
> @Arthur-GYT @biodranik should the new type be validated in the `indexer/editable_map_object.cpp`? :
> 
> ```c++
> bool EditableMapObject::IsValidMetadata(MetadataID type, std::string const & val
> {
>   switch (type)
>   {
>   case MetadataID::FMD_WEBSITE: return ValidateWebsite(value);
>   case MetadataID::FMD_WEBSITE_MENU: return ValidateWebsite(value); // <--- ?
> 
> void EditableMapObject::SetMetadata(MetadataID type, std::string value)
> {
>   switch (type)
>   {
>   case MetadataID::FMD_WEBSITE: value = ValidateAndFormat_website(value); break;
>   case MetadataID::FMD_WEBSITE_MENU: value = ValidateAndFormat_website(value); break;  // <--- ?
> ...
> ```

I forgot this ^^'. It's added, thanks ^^.

**Comment by kirylkaveryn** [comment]
@Arthur-GYT also important to add the new key into the `editor/editor_config.cpp` to te list of fields with metadata. In other case you will get a crash while parsing the `editor.config` XML tags.
```
   {"operator", EType::FMD_OPERATOR},
    {"website", EType::FMD_WEBSITE},
    {"website_menu", EType::FMD_WEBSITE_MENU}, // <---
    {"contact_facebook", EType::FMD_CONTACT_FACEBOOK},
```
And only after that, the `editor.config` can be updated to support a new tag.
@biodranik @dvdmrtnz can you please correct me if I'm wrong?

**Comment by kirylkaveryn** [comment]
> 3\. Both cuisine and menu should be nearby in POI (they are closely related) and in the editor (we really want to encourage users to add menus, not wifis or something else, this is very valuable data for planning a visit to a restaurant).

@biodranik 
The list of editable objects is autogenerated and the FMD_WEBSITE_MENU is the last in the types enum (46) so in ios it is always appended to the end... I found that we don't reorder the list manually In ios.
image

Simulator Screen Recording - iPhone 15 Pro - 2024-05-12 at 17 28 09

Simulator Screen Recording - iPhone 15 Pro - 2024-05-12 at 17 27 39

**Comment by Arthur-GYT** [comment]
@kirylkaveryn It's changed ^^

**Comment by biodranik** [comment]
It should be possible to change the order of the visible fields on the screen, the internal order can be anything unrelated.

**Comment by kirylkaveryn** [comment]
> It should be possible to change the order of the visible fields on the screen, the internal order can be anything unrelated.

It requires a refactroing and may be implemented in a separate PR...

**Comment by biodranik** [comment]
> It requires refactoring and may be implemented in a separate PR...

A separate PR is a good idea!

**Review by biodranik — APPROVED** [review]
Ready to merge?

**Comment by Arthur-GYT** [comment]
For me, yes

**Review by vng — APPROVED** [review]

**Comment by biodranik** [comment]
@Arthur-GYT congrats and thanks for your hard work and contribution!

**Pull request merged**
